### DISCUSSION TITLE
Whitespace tokenization

### DISCUSSION THREAD

**Discussion by Praful932** [explosion/spaCy#7271] [Help: Coding & Implementations] [answered]
Hi, I have testing data and annotations which I cannot modify. And as a tokenization step, I'm using the spacy tokenizer. The text data also contains more than one whitespace in some sentences. Is it possible to modify the tokenizer to skip the whitespaces which occur during tokenization. Example below
```python
import spacy
from spacy.lang.en import English
nlp2 = English()
s = 'ABCD:  lorem ipsum dolor a)b'
for t in nlp2.tokenizer(s):
  print([t.text])
```
`> ['ABCD']
[':']
[' ']
['lorem']
['ipsum']
['dolor']
['a)b']`

Thanks

**Comment by polm** [comment, marked as answer]
Dealing with multiple spaces is always kind of awkward. Because spaCy makes a point of preserving the original input, and because the `whitespace_` attribute is special and can only return a single space, multiple spaces result in space tokens like this. Tabs, newlines, and more exotic spaces will also get their own tokens.

Space tokens like this have an `is_space` attribute you can use to filter them out before passing output to downstream tasks. Is using that enough or is there another reason the existence of the tokens in the Doc is causing you issues?

**Reply by Praful932** [reply]
Thanks @polm for the response. For my use case, I cannot shift the original data which has annotations(that would change too) and also apart from whitespace there were a lot of edge cases since I was using the token.idx later on to map back to the original data. So To get my result I coded up a workaround which is an inffecient string match 😅, it's working for now

**Reply by polm** [reply]
Glad to hear you found a working solution. I'm still a little unclear on what problem the space tokens are giving you exactly though - do you need `lorem` in your example to be the third token for some reason, for example? Can you give an example of your annotation format, maybe that'll make it clear?

**Reply by Praful932** [reply]
So it is a NER problem, the raw_data has annotations character indexed, there can be partial/full overlap annotations too. So down the pipeline I'm using HuggingFace transformer which needs token based ground truth labels, so for that I'm using a modified spacy tokenizer to first tokenize the text and then using some code to convert the char-based annotations to token based. Since HF uses a BPE tokenizer to train. After training I need to convert the token based predictions back to char indexed true format. So till here I did not face any issue, the problem comes when I want to map the predictions back, so for eg if the BPE split the token `'ABCD'` to `'abc'` & `'##d'` I can map that back using an attribute of the HF tokenizer, although this isn't possible if there is a token which has been tokenized before in the pipeline by spacy but skipped by Hugging face in this case `'  '`. So since I want to map it back to the original spacy tokenized, this gave an issue(other edge cases were there too). So currently I'm scanning the non-tokenized with the attribute that HF gives to map back to the tokens variable and appending the annotations accordingly 😅
Here's an example
```python
tokens = []
s = 'ABCD:  lorem ipsum dolor a)b'
for token in nlp.tokenizer(s):
  tokens.append(token.text)
print(tokens)
print()
for token in tokenizer.tokenize(tokens, truncation = True, is_split_into_words = True, padding = True, max_length = max_length):
  print([token])
```
```
['ABCD', ':', ' ', 'lorem', 'ipsum', 'dolor', 'a', ')', 'b']

['abc']
['##d']
[':']
['lor']
['##em']
['ips']
['##um']
['dol']
['##or']
['a']
[')']
['b']
```

**Reply by adrianeboyd** [reply]
If it's helpful, we use this library in `spacy-transformers` to do alignments between spacy tokens and transformer tokens:

https://github.com/explosion/spacy-alignments

We would like to switch to the alignments provided by `tokenizers` at some point, but the last time I checked it wasn't working 100% for all the versions / model types we want to support.

**Reply by Praful932** [reply]
This is nice, Thanks, I'll check it out
